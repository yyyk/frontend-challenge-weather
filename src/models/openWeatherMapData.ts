export interface OpenWeatherMapData {
  city: {
    coord: {
      lat: number;
      lon: number;
    };
    country: string;
    id: number;
    name: string;
  };
  cnt: number;
  cod: string;
  list: Array<{
    clouds: {
      all: number;
    };
    dt: number;
    dt_txt: string;
    main: {
      grnd_level: number;
      humidity: number;
      pressure: number;
      sea_level: number;
      temp: number; // in Kelvin unit
      temp_kf: number;
      temp_max: number; // in Kelvin unit
      temp_min: number; // in Kelvin unit
    };
    sys: {
      pod: string;
    };
    weather: Array<{
      description: string;
      icon: string;
      id: number;
      main: string;
    }>;
    wind: {
      deg: number;
      speed: number;
    };
  }>;
}
