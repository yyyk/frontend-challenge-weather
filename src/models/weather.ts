import { City } from './city';

export type WeatherType = 'Clear' | 'Clouds';

export interface Weather {
  hour: number;
  type: WeatherType;
  temperature: number; // in Celsius unit
}

export interface WeatherData {
  city: City;
  dateText: string;
  day: number;
  date: number;
  month: number;
  maxTemperature: number; // in Celsius unit
  minTemperature: number; // in Celsius unit
  weatherList: Weather[];
}
