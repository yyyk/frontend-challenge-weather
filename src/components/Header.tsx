import React, { FC } from 'react';
import styled from '@emotion/styled';
import { DAY_LIST } from 'src/constants/dayList';
import { MONTH_LIST } from 'src/constants/monthList';
import colors from 'src/constants/colors';
import { WeatherData, WeatherType } from 'src/models/weather';
import { fluidType } from 'src/utils/fluidType';
import { useSelectedWeather } from 'src/providers/SelectedWeather';
import Sun from './icons/Sun';
import Cloud from './icons/Cloud';

export interface HeaderProps {
  weatherData: WeatherData;
}

const Container = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-evenly;
  align-items: flex-start;
  background: ${colors.background.darkBlue};
`;

const SVGIconContainer = styled.div`
  width: 17%;
  height: auto;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  justify-content: center;
  align-self: stretch;
`;

const TemperatureContainer = styled.div`
  width: 23.33%;
`;

const TemperatureText = styled.div`
  font-style: normal;
  font-weight: bold;
  color: ${colors.white};
  font-size: ${fluidType(300, 1920, 38, 246)};
  line-height: ${fluidType(300, 1920, 40, 254)};
  text-align: center;
`;

const SubText = styled.div`
  font-style: normal;
  font-weight: normal;
  font-size: ${fluidType(300, 1920, 8, 48)};
  line-height: ${fluidType(300, 1920, 10, 48)};
  color: ${colors.text.gray};
  margin: 0 0 1.6vw;
`;

const TemperatureInfoText = styled(SubText)`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
`;

const DateContainer = styled.div`
  width: auto;
`;

const CityNameText = styled(SubText)``;

const DateText = styled.div`
  font-style: normal;
  font-weight: bold;
  font-size: ${fluidType(300, 1920, 15, 99)};
  line-height: ${fluidType(300, 1920, 20, 124)};
  color: ${colors.white};
`;

const Header: FC<HeaderProps> = ({ weatherData }) => {
  const { selectedWeather } = useSelectedWeather();
  const renderIcon = (type: WeatherType) => {
    switch (type) {
      case 'Clear':
        return <Sun />;
      case 'Clouds':
        return <Cloud />;
      default:
        return null;
    }
  };
  return (
    <Container>
      <SVGIconContainer>{selectedWeather && renderIcon(selectedWeather?.type)}</SVGIconContainer>
      <TemperatureContainer>
        <TemperatureInfoText>
          <div>{selectedWeather?.type}</div>
          <div>
            {weatherData.maxTemperature}&deg; / {weatherData.minTemperature}&deg;
          </div>
        </TemperatureInfoText>
        <TemperatureText>{selectedWeather?.temperature}&deg;</TemperatureText>
      </TemperatureContainer>
      <DateContainer>
        <CityNameText>{weatherData?.city?.name}</CityNameText>
        <DateText>{DAY_LIST[weatherData?.day]}</DateText>
        <DateText>
          {weatherData.date}. {MONTH_LIST[weatherData.month]}
        </DateText>
      </DateContainer>
    </Container>
  );
};

export default Header;
