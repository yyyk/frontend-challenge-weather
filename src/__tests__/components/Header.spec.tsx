import { render, screen } from '@testing-library/react';
import { Weather, WeatherData } from 'src/models/weather';
import MOCK_DATA from 'src/__tests__/mock/weatherData.json';
import SelectedWeatherProvider from 'src/providers/SelectedWeather';
import Header from 'src/components/Header';

describe('<Header>', () => {
  let weatherData: WeatherData;
  let selectedWeather: Weather;

  beforeEach(() => {
    weatherData = MOCK_DATA as WeatherData;
    selectedWeather = {
      hour: 3,
      type: 'Clear',
      temperature: -5,
    };
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Header weatherData={weatherData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Sun icon', async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('Clear', { selector: 'title' })).toBeTruthy();
  });

  it('renders Cloudy icon', async () => {
    selectedWeather = {
      hour: 3,
      type: 'Clouds',
      temperature: -5,
    };
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('Cloudy', { selector: 'title' })).toBeTruthy();
  });

  it('renders weather text', async () => {
    selectedWeather = {
      hour: 3,
      type: 'Clouds',
      temperature: -5,
    };
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('Clouds')).toBeTruthy();
  });

  it('renders max temperature and minimum temperature', async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('6° / -5°')).toBeTruthy();
  });

  it("renders selected hour's temperature", async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('-5°')).toBeTruthy();
  });

  it('renders city name', async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('Munich')).toBeTruthy();
  });

  it('renders day', async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('Monday')).toBeTruthy();
  });

  it('renders date and month', async () => {
    await render(
      <SelectedWeatherProvider initialState={selectedWeather}>
        <Header weatherData={weatherData} />
      </SelectedWeatherProvider>,
    );
    expect(screen.getByText('20. February')).toBeTruthy();
  });
});
