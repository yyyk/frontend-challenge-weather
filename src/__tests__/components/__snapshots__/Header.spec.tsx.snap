// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-18eox44"
  >
    <div
      class="css-2vaek4"
    />
    <div
      class="css-1pkeeh3"
    >
      <div
        class="css-1j8xtbb"
      >
        <div />
        <div>
          6° / -5°
        </div>
      </div>
      <div
        class="css-5thqwk"
      >
        °
      </div>
    </div>
    <div
      class="css-9msay7"
    >
      <div
        class="css-jooq6e"
      >
        Munich
      </div>
      <div
        class="css-115lm8a"
      >
        Monday
      </div>
      <div
        class="css-115lm8a"
      >
        20. February
      </div>
    </div>
  </div>
</DocumentFragment>
`;
